src/transform/src/attribute/mod.rs

```diff
@@ -154,3 +172,200 @@ enum EnvTask<T> {
     NoOp,
 }
+
+#[allow(missing_debug_implementations)]
+#[derive(Default)]
+/// Builds an [AttributeDeriver]
+pub struct AttributeBuilder {
+    attributes: TypeMap,
+}
+
+impl AttributeBuilder {
+    /// Add an attribute that [AttributeDeriver] should derive
+    pub fn add_attribute<T: TypeMapKey>(&mut self)
+    where
+        T::Value: Attribute + Default,
+    {
+        if !self.attributes.contains_key::<T>() {
+            T::Value::add_dependencies(self);
+            self.attributes.insert::<T>(T::Value::default());
+        }
+    }
+
+    /// Consume `self`, producing an [AttributeDeriver]
+    pub fn finish(self) -> AttributeDeriver {
+        AttributeDeriver {
+            attributes: self.attributes,
+        }
+    }
+}
+
+#[allow(missing_debug_implementations)]
+/// Derives an attribute and any attribute it depends on.
+///
+/// Can only be constructed from a [AttributeBuilder].
+pub struct AttributeDeriver {
+    attributes: TypeMap,
+}
+
+/// A topological sort of extant attributes.
+///
+/// The attribute assigned value 0 depends on no other attributes.
+/// We expect the attributes to be assigned numbers in the range
+/// 0..TOTAL_ATTRIBUTES with no gaps in between.
+#[derive(FromPrimitive)]
+enum AttributeId {
+    SubtreeSize = 0,
+    NonNegative = 1,
+    Arity = 2,
+    RelationType = 3,
+    UniqueKeys = 4,
+}
+
+/// Should always be equal to the number of attributes
+const TOTAL_ATTRIBUTES: usize = 5;
+
+impl AttributeDeriver {
+    /// Does derivation work required upon entering a subexpression
+    pub fn pre_visit(&mut self, expr: &MirRelationExpr) {
+        fn schedule_env_tasks<T: TypeMapKey>(attributes: &mut TypeMap, expr: &MirRelationExpr)
+        where
+            T::Value: Attribute,
+        {
+            if let Some(attr) = attributes.get_mut::<T>() {
+                attr.schedule_env_tasks(expr)
+            }
+        }
+        for i in 0..TOTAL_ATTRIBUTES {
+            match FromPrimitive::from_usize(i) {
+                Some(AttributeId::SubtreeSize) => {
+                    schedule_env_tasks::<SubtreeSize>(&mut self.attributes, expr);
+                }
+                Some(AttributeId::NonNegative) => {
+                    schedule_env_tasks::<NonNegative>(&mut self.attributes, expr);
+                }
+                Some(AttributeId::RelationType) => {
+                    schedule_env_tasks::<RelationType>(&mut self.attributes, expr);
+                }
+                Some(AttributeId::Arity) => {
+                    schedule_env_tasks::<Arity>(&mut self.attributes, expr);
+                }
+                Some(AttributeId::UniqueKeys) => {
+                    schedule_env_tasks::<UniqueKeys>(&mut self.attributes, expr);
+                }
+                None => {}
+            }
+        }
+    }
+
+    /// Does derivation work required upon exiting a subexpression
+    pub fn post_visit(&mut self, expr: &MirRelationExpr) {
+        let mut deps = TypeMap::default();
```

Comment:
FYI: I think that you only need to allocate two maps for the entire derivation process, but I need to play with the code in order to suggest something more concrete. Let's merge it like that for now with an eye on optimizing these allocations as a follow-up.